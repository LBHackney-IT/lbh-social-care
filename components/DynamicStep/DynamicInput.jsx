import PropTypes from 'prop-types';

import * as Inputs from 'components/Form';

const DynamicInput = ({
  component,
  register,
  control,
  id,
  name,
  errors,
  multiStepIndex,
  ...otherProps
}) => {
  const inputName =
    typeof multiStepIndex === 'number'
      ? `${id}[${multiStepIndex}].${name}`
      : name;
  const Component = Inputs[component];
  const sharedProps = {
    name: inputName,
    error: errors[inputName],
    required: otherProps?.rules?.required,
    ...otherProps,
  };
  switch (component) {
    case 'AddressLookup':
    case 'DateInput':
      return <Component control={control} {...sharedProps} />;
    default:
      return (
        <Component register={register(otherProps.rules)} {...sharedProps} />
      );
  }
};

DynamicInput.propTypes = {
  component: PropTypes.string.isRequired,
  register: PropTypes.func.isRequired,
  control: PropTypes.object.isRequired,
  id: PropTypes.string.isRequired,
  name: PropTypes.string.isRequired,
  errors: PropTypes.object.isRequired,
  rules: PropTypes.shape({
    required: PropTypes.oneOfType([PropTypes.bool, PropTypes.string]),
  }),
  multiStepIndex: PropTypes.number,
};

export default DynamicInput;
