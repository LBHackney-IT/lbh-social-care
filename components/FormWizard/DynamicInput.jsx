import PropTypes from 'prop-types';
import get from 'lodash/get';

import * as Inputs from 'components/Form';

const enhanceValidate = (validate, currentData) =>
  typeof validate === 'function'
    ? (value) => validate(value, currentData)
    : Object.entries(validate).reduce(
        (acc, [key, validation]) => ({
          ...acc,
          [key]: (value) => validation(value, currentData),
        }),
        {}
      );

const DynamicInput = (props) => {
  const {
    component,
    control,
    errors,
    name,
    options,
    currentData,
    register,
    rules,
    ...otherProps
  } = props;
  const Component = Inputs[component];
  if (!Component)
    throw new Error(`"${component}" is not a supported component type.`);
  const sharedProps = {
    name,
    error: get(errors, name),
    required: rules?.required,
    options: typeof options === 'function' ? options(currentData) : options,
    rules: {
      ...rules,
      validate: rules?.validate && enhanceValidate(rules.validate, currentData),
    },
    ...otherProps,
  };
  if (typeof options === 'function' && !sharedProps.options) {
    return null;
  }
  switch (component) {
    case 'ObjectInput':
      return <Component {...props} />;
    case 'AddressLookup':
    case 'Autocomplete':
    case 'DateInput':
      return <Component control={control} {...sharedProps} />;
    default:
      return <Component register={register} {...sharedProps} />;
  }
};

DynamicInput.propTypes = {
  component: PropTypes.string.isRequired,
  control: PropTypes.object.isRequired,
  name: PropTypes.string.isRequired,
  errors: PropTypes.object.isRequired,
  currentData: PropTypes.object.isRequired,
  labelSize: PropTypes.string,
  rules: PropTypes.shape(),
  register: PropTypes.func,
  options: PropTypes.oneOfType([PropTypes.func, PropTypes.array]),
};

export default DynamicInput;
