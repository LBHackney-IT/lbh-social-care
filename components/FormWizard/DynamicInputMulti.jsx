import { useState, useCallback } from 'react';
import PropTypes from 'prop-types';
import cx from 'classnames';

import DynamicInput from 'components/FormWizard/DynamicInput';
import DeleteIcon from 'components/Icons/TimesCircle';

import styles from './DynamicInputMulti.module.scss';

const DynamicInputMulti = ({
  name,
  initialInputData = [],
  currentData,
  onDelete,
  label,
  hint,
  isMultiTrigger = 'Add a new one',
  isMultiInit = true,
  ...otherProps
}) => {
  const multiPresent = initialInputData.length || 0;
  const [counter, setCounter] = useState(
    multiPresent === 0 && isMultiInit ? 1 : 0
  );
  const removeSelected = useCallback((index) => {
    setCounter(counter - 1);
    currentData[name] &&
      onDelete([
        ...currentData[name].slice(0, index),
        ...currentData[name].slice(index + 1, currentData[name].length),
      ]);
  });
  return (
    <>
      {Array.apply(null, { length: counter + multiPresent }).map((e, index) => (
        <div
          key={`${name}${index}`}
          className={cx('govuk-!-margin-bottom-5', styles.wrapper)}
        >
          <DynamicInput
            {...otherProps}
            name={`${name}[${index}]`}
            currentData={currentData}
            label={index === 0 ? label : null}
            hint={index === 0 ? hint : null}
          />
          {(!isMultiInit || index !== 0) && (
            <span
              className={styles.delete}
              role="button"
              onClick={() => removeSelected(index)}
            >
              <DeleteIcon />
            </span>
          )}
        </div>
      ))}
      <div className="govuk-!-margin-top-3 govuk-!-margin-bottom-5">
        <span
          className="govuk-link"
          role="button"
          onClick={() => setCounter(counter + 1)}
        >
          {isMultiTrigger}
        </span>
      </div>
    </>
  );
};

DynamicInputMulti.propTypes = {
  name: PropTypes.string.isRequired,
  label: PropTypes.string,
  hint: PropTypes.string,
  initialInputData: PropTypes.array,
  currentData: PropTypes.object.isRequired,
  onDelete: PropTypes.func.isRequired,
  isMultiTrigger: PropTypes.string,
};

export default DynamicInputMulti;
