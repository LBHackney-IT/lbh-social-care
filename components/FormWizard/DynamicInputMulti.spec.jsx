import { render, fireEvent } from '@testing-library/react';

import DynamicInputMulti from './DynamicInputMulti';

import DynamicInput from 'components/FormWizard/DynamicInput';

jest.mock('components/FormWizard/DynamicInput', () =>
  jest.fn(() => 'MockedDynamicInput')
);

jest.mock('components/Icons/TimesCircle', () => () => 'MockedCloseButton');

describe('DynamicInputMulti', () => {
  const props = {
    label: 'i am a multi input',
    hint: 'i accept multiple inputs',
    name: 'dynamic',
    initialInputData: ['f', 'o', 'o'],
    currentData: { foo: 'bar' },
    isMulti: true,
    component: 'Dynamic',
    foo: 'bar',
    onDelete: jest.fn(),
  };

  it('should render properly', () => {
    const { asFragment } = render(<DynamicInputMulti {...props} />);
    expect(asFragment()).toMatchSnapshot();
  });

  it('should render properly with isMultiInit false', () => {
    const newProps = {
      isMultiInit: false,
      initialInputData: [],
      isMultiTrigger: 'Add a something',
    };
    const { asFragment } = render(
      <DynamicInputMulti {...props} {...newProps} />
    );
    expect(asFragment()).toMatchSnapshot();
    expect(DynamicInput).not.toHaveBeenCalled();
  });

  it('should pass the correct props to DynamicInput', () => {
    render(<DynamicInputMulti {...props} />);
    expect(DynamicInput).toHaveBeenCalled();
    expect(DynamicInput).toHaveBeenNthCalledWith(
      1,
      {
        label: 'i am a multi input',
        hint: 'i accept multiple inputs',
        name: 'dynamic[0]',
        currentData: { foo: 'bar' },
        isMulti: true,
        component: 'Dynamic',
        foo: 'bar',
      },
      {}
    );
    expect(DynamicInput).toHaveBeenNthCalledWith(
      2,
      {
        label: null,
        hint: null,
        name: 'dynamic[1]',
        currentData: { foo: 'bar' },
        isMulti: true,
        component: 'Dynamic',
        foo: 'bar',
      },
      {}
    );
    expect(DynamicInput).toHaveBeenNthCalledWith(
      3,
      {
        label: null,
        hint: null,
        name: 'dynamic[2]',
        currentData: { foo: 'bar' },
        isMulti: true,
        component: 'Dynamic',
        foo: 'bar',
      },
      {}
    );
  });

  it('should add a new input on "Add a new one"', () => {
    const { getByText, getAllByText } = render(
      <DynamicInputMulti
        {...props}
        initialInputData={['foo', 'bar']}
        currentData={{ dynamic: [1, 2, 3] }}
      />
    );
    expect(getAllByText('MockedCloseButton').length).toBe(1);
    fireEvent.click(getByText('Add a new one'));
    expect(getAllByText('MockedCloseButton').length).toBe(2);
  });

  it('should pass the updated data to onDelete', () => {
    const { getByText } = render(
      <DynamicInputMulti
        {...props}
        initialInputData={['foo', 'bar']}
        currentData={{ dynamic: [1, 2, 3] }}
      />
    );
    fireEvent.click(getByText('MockedCloseButton'));
    expect(props.onDelete).toHaveBeenCalled();
    expect(props.onDelete).toHaveBeenCalledWith([1, 3]);
  });
});
