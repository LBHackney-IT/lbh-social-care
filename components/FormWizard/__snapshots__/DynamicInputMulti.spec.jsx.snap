// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`DynamicInputMulti should render properly 1`] = `
<DocumentFragment>
  <div
    class="govuk-!-margin-bottom-5 wrapper"
  >
    MockedDynamicInput
  </div>
  <div
    class="govuk-!-margin-bottom-5 wrapper"
  >
    MockedDynamicInput
    <button
      class="delete"
    >
      MockedCloseButton
      <span
        class="govuk-visually-hidden"
      >
        Remove this i am a multi input
      </span>
    </button>
  </div>
  <div
    class="govuk-!-margin-bottom-5 wrapper"
  >
    MockedDynamicInput
    <button
      class="delete"
    >
      MockedCloseButton
      <span
        class="govuk-visually-hidden"
      >
        Remove this i am a multi input
      </span>
    </button>
  </div>
  <div
    class="govuk-!-margin-bottom-5"
  >
    <button
      class="govuk-button lbh-button lbh-button--secondary lbh-button--add govuk-!-margin-top-0"
      role="button"
    >
      <svg
        fill="none"
        height="11"
        viewBox="0 0 11 11"
        width="11"
        xmlns="http://www.w3.org/2000/svg"
      >
        <rect
          fill="#00664F"
          height="1.94118"
          transform="rotate(-90 4.5293 11)"
          width="11"
          x="4.5293"
          y="11"
        />
        <rect
          fill="#00664F"
          height="1.94118"
          width="11"
          y="4.52942"
        />
      </svg>
      Add another
    </button>
  </div>
</DocumentFragment>
`;

exports[`DynamicInputMulti should render properly with isMultiInit false 1`] = `
<DocumentFragment>
  <div
    class="govuk-!-margin-bottom-5"
  >
    <button
      class="govuk-button lbh-button lbh-button--secondary lbh-button--add govuk-!-margin-top-0"
      role="button"
    >
      <svg
        fill="none"
        height="11"
        viewBox="0 0 11 11"
        width="11"
        xmlns="http://www.w3.org/2000/svg"
      >
        <rect
          fill="#00664F"
          height="1.94118"
          transform="rotate(-90 4.5293 11)"
          width="11"
          x="4.5293"
          y="11"
        />
        <rect
          fill="#00664F"
          height="1.94118"
          width="11"
          y="4.52942"
        />
      </svg>
      Add a something
    </button>
  </div>
</DocumentFragment>
`;
